python/langsmith/_internal/_background_thread.py

```diff
@@ -88,6 +94,42 @@ def _tracing_thread_drain_queue(
     return next_batch
 
 
+def _tracing_thread_drain_compressed_buffer(
+    client: Client,
+    size_limit: int = 100,
+    size_limit_bytes: int = 50 * 1024 * 1024
+) -> Optional[Iterable[bytes]]:
+    with client._buffer_lock:
+        current_size = client.compressed_runs_buffer.tell()
+
+        # Check if we should send now
+        if not (client._run_count >= size_limit or current_size >= size_limit_bytes):
+            return None
+
+        # Write final boundary and close compression stream
+        client.compressor_writer.write(f'--{client.boundary}--\r\n'.encode())
+        client.compressor_writer.flush()
+        client.compressor_writer.close()
+
+        client.compressed_runs_buffer.seek(0)
+
+        def data_stream() -> Iterable[bytes]:
+            chunk_size = 65536
+            while True:
+                chunk = client.compressed_runs_buffer.read(chunk_size)
+                if not chunk:
+                    break
+                yield chunk
+
+        # Reinitialize for next batch
+        client.compressed_runs_buffer = io.BytesIO()
+        client.compressor = zstd.ZstdCompressor()
+        client.compressor_writer = client.compressor.stream_writer(
+            client.compressed_runs_buffer, closefd=False)
+        client._run_count = 0
+
+        return data_stream()
```

Comment:
this doesn't look right to me. You are returning `data_stream` which operates on the buffer you just reset?

I'd move the existing buffer to a separate variable as soon you finish closing, then reset the buffer, then release the lock

Reply:
yeah good catch, missed this when implementing streaming